Binary patching of build prefixes

danielnachun opened the pull request:
Now that we've merged https://github.com/Homebrew/brew/pull/12890, we can narrow the focus of this PR only to what is necessary to enable binary patching.  I've written a new unit test `keg_relocate/binary_relocation_spec.rb` to test `relocate_build_prefix`, which is by far the most complicated part of the PR.  This should help us avoid breaking this method in the future.   I'm not sure how practical it is implement useful unit tests for the other changes but they're not as complex.

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,49 @@ def replace_text_in_files(relocation, files: nil)
+    end
+
+    binary_string_matches = Set.new
+    keg.each_unique_binary_file do |file|
+      binary_string_matches << file if string_matches.include?(file)
+    end
```
Might need a refactoring of `each_unique_binary_file` but it looks like you could make this only read the `string_matches` files if you're only adding if matches are already in there.

danielnachun replied:
I tested this informally by calling `egrep -rlO "\x00" $(fgrep -rlO "/usr/local")` and it is noticeably faster.  I think all I need to do is add an optional `files` argument to `each_unique_binary_file` (similar to what we do for some other methods) which just passes the file list to `egrep` so it only scans those files instead of the whole keg.  I'll have to slightly tweak the unit test as well but it will be simple enough.

danielnachun replied:
This ended up being even simpler than I thought.  I just added `files` as an argument to `each_unique_binary_file` and initialized it to `nil`. If it's empty, `egrep` will just scan all files on the keg.  If it's not empty, only the files passed to it will be scanned, which is now what happens in `relocate_build_prefix`.  Testing this in some formulae with a large number of files, this does lead to a substantial speedup so this was a good improvement.

MikeMcQuaid replied:
Makes sense to me!

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,49 @@ def replace_text_in_files(relocation, files: nil)
+
+    # Only consider string matches which are binary files with null bytes, and remove any
+    # matches which are sharballs found by text_files.
+    binary_string_matches -= text_files
```
Similarly, rather than removing these files after reading them: consider not reading them at all.

danielnachun replied:
Yes, I should just remove `text_files` from `string_matches` earlier.  Combined with the above change, that will mean a much higher proportion of the files being read by `each_unique_binary` are likely to be binary files.

danielnachun replied:
Done.

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,49 @@ def replace_text_in_files(relocation, files: nil)
+    binary_string_matches.each do |binary_file|
+      binary_file.ensure_writable do
+        binary = File.binread binary_file
+        puts "Replacing build prefix in: #{binary_file}"
```
This should probably be `odebug` or require `--verbose` or `HOMEBREW_DEVELOPER`.

danielnachun replied:
I gated this behind `verbose` as I do think it's useful in kegs with a lot of files that need patching to be able to see this progress and know that replacement is happening successfully, but should not displayed by default.

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,49 @@ def replace_text_in_files(relocation, files: nil)
+        # Only perform substitution on strings which match prefix regex.
+        match_indices.each do |i|
+          s = binary_strings[i]
+          binary_strings[i] = s.gsub(old_prefix, new_prefix).ljust(s.size, NULL_BYTE)
```
```suggestion
          binary_strings[i] = s.sub(old_prefix, new_prefix)
                               .ljust(s.size, NULL_BYTE)
```

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,49 @@ def replace_text_in_files(relocation, files: nil)
+
+        # Add back null padding at the end of the binary if needed.
+        patched_binary = binary_strings.join(NULL_BYTE).ljust(binary.size, NULL_BYTE)
+        if patched_binary.size != binary.size
+          raise "Patching failed!  Original and patched binary sizes do not match."
+        end
```
```suggestion
        if patched_binary.size != binary.size
          raise <<~EOS
            Patching failed! Original and patched binary sizes do not match.
            Original size: #{binary.size}
            Patched size:  #{patched_binary.size}
          EOS
        end
```

carlocab commented on Library/Homebrew/software_spec.rb:
```diff
@@ -513,7 +520,30 @@ def compatible_locations?(tag: Utils::Bottles.tag)
+    compatible_cellar && compatible_prefix
+  end
+
+  # Should the build prefix for the {Bottle} this {BottleSpecification} belongs to be relocated?
```
Phrasing the docstring as a question here is a little weird.

MikeMcQuaid replied:
Agreed.

carlocab commented on Library/Homebrew/software_spec.rb:
```diff
@@ -513,7 +520,30 @@ def compatible_locations?(tag: Utils::Bottles.tag)
+
+  # Should the build prefix for the {Bottle} this {BottleSpecification} belongs to be relocated?
+  sig { params(tag: Utils::Bottles::Tag).returns(T::Boolean) }
+  def skip_prefix_relocation?(tag: Utils::Bottles.tag)
```
This method seems to have a lot of overlap with `#compatible_locations?`. I wonder if they really need to be separate methods.

MikeMcQuaid replied:
Agreed.

danielnachun replied:
I had to think the about this a bit but I think the proper solution here is to have a `cellar_from_tag` method which returns the cellar.  Then both times where we need to know the bottle cellar, we can just get it from the tag.

danielnachun commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,57 @@ def replace_text_in_files(relocation, files: nil)
+    binary_string_matches.each do |binary_file|
+      binary_file.ensure_writable do
+        binary = File.binread binary_file
+        puts "Replacing build prefix in: #{binary_file}" if verbose
```
That does sound cleaner, I'll try that instead.

MikeMcQuaid replied:
Agreed.

danielnachun commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -203,17 +254,21 @@ def each_unique_file_matching(string, &block)
     end
   end
 
+  def codesign_patched_binary(_binary_file)
+    []
+  end
```
That make more sense.  I suppose this was the first instance where `apply_ad_hoc_signature` is being called from OS-generic code so this was unneeded before.

MikeMcQuaid replied:
Can also consider renaming this method if desired.

danielnachun commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,57 @@ def replace_text_in_files(relocation, files: nil)
+                               .ljust(s.size, NULL_BYTE)
+        end
+
+        # Add back null padding at the end of the binary if needed.
+        patched_binary = binary_strings.join(NULL_BYTE).ljust(binary.size, NULL_BYTE)
```
The reason I had to do this was because by default, `.split` truncates trailing empty strings.  So if the last few bytes of the binary are null bytes (i.e. `some_string\x00\x00\x00\x00`, then the last strings in `binary_strings` will be empty and be dropped.  However, I see here that if I were to set the optional `limit` parameter of `.split` to `-1`, the empty strings are not dropped:  https://www.geeksforgeeks.org/ruby-string-split-method-with-examples/.  I was not aware of this before but I think it's worth trying to implement, because if it works we could get rid the second `ljust` entirely, making this much simpler.

carlocab replied:
Can we rule out leading null bytes? If not, then we may want to split on a regex instead of just a raw null byte. See https://ruby-doc.org/core-2.6.3/String.html#method-i-split

danielnachun replied:
I believe leading null bytes don't normally occur in ELF/MachO binaries, but this may not be true in other binary file types so I don't know that we can rule them out.  From my current testing, it seems that using a regex works just as well and may be safer, though I still have to add a limit of -1 to avoid dropping the trailing empty strings.

MikeMcQuaid commented on Library/Homebrew/extend/os/mac/keg_relocate.rb:
```diff
@@ -189,4 +189,8 @@ def egrep_args
     [grep_bin, grep_args]
   end
+
+  def codesign_patched_binary(binary_file)
+    apply_ad_hoc_signature(binary_file)
+  end
```
Yeh, that would work for me. Could rename that method if we prefer this naming (I kinda do).

MikeMcQuaid commented on Library/Homebrew/software_spec.rb:
```diff
@@ -513,7 +520,30 @@ def compatible_locations?(tag: Utils::Bottles.tag)
-    return true if [:any, :any_skip_relocation].include?(cellar)
+    return true if RELOCATABLE_CELLARS.include?(cellar)
+
+    prefix = Pathname(cellar).parent.to_s
```
The lack of any test coverage for much of this file worries me a little.

MikeMcQuaid commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,57 @@ def replace_text_in_files(relocation, files: nil)
 
+  def relocate_build_prefix(keg, old_prefix, new_prefix, verbose)
+    # Find binaries which match prefix strings.
+    string_matches = Set.new
+    keg.each_unique_file_matching(old_prefix) do |file|
+      string_matches << file
+    end
+
+    # Remove files that are known to be plain text.
+    # Note that this also removes sharballs, which we currently don't patch
+    # because they appear to break after patching.
+    string_matches -= text_files
+
+    # Only keep files with prefix matches which contain null bytes.  We need to do this
+    # because text_files is not intended to identify all types of text files.
+    binary_string_matches = Set.new
+    keg.each_unique_binary_file(string_matches) do |file|
+      binary_string_matches << file
+    end
```
I'm not convinced fewer loops is necessarily better here. I think it's worth minimising file reads but not loops. As far as I understand now, each loop is through a subset of the previous one. If there's a way to combine reads more: I'd say go for it but I think combining loops may harm readability.

carlocab replied:
I agree re minimising file reads instead of loops, but I would also say doing all three loops in one instead is actually much better for readability than otherwise.

danielnachun replied:
While it is true that each loop iterates over a subset of the previous one, I do kind of feel like that structure is leftover boilerplate from when I was using a much less effective way of identifying binary files.  Now that we have a very simple strategy for doing this, I am inclined to have it look something like this:
```
text_files_list = text_files # store output of text_files since it has to read all files in keg

keg.each_unique_file_matching(old_prefix) do |file|
   next if text_files_list.include?(file) # skip files we know are plain text
   next if !file.binary_file? # skip files which do not contain null bytes

   # binary patching code
end
```
I think this would actually be easier to read than how it's currently written.  I can push a separate commit implementing this to see how it looks and we can decide if the readability has improved or worsened.

carlocab replied:
That's pretty much exactly what I had in mind.

Also, I wonder if we should be doing the caching in `#text_files` with something like
```diff
diff --git a/Library/Homebrew/keg_relocate.rb b/Library/Homebrew/keg_relocate.rb
index ad779064b..fd0332f15 100644
--- a/Library/Homebrew/keg_relocate.rb
+++ b/Library/Homebrew/keg_relocate.rb
@@ -223,8 +223,9 @@ class Keg
   end
 
   def text_files
-    text_files = []
-    return text_files if !which("file") || !which("xargs")
+    @text_files ||= []
+    return @text_files if !which("file") || !which("xargs")
+    return @text_files unless @text_files.blank?
 
     # file has known issues with reading files on other locales. Has
     # been fixed upstream for some time, but a sufficiently new enough
@@ -239,7 +240,7 @@ class Keg
         next true if Metafiles::EXTENSIONS.include?(pn.extname)
 
         if pn.text_executable?
-          text_files << pn
+          @text_files << pn
           next true
         end
         false
@@ -260,11 +261,11 @@ class Keg
         path = Pathname.new(path)
         next unless files.include?(path)
 
-        text_files << path
+        @text_files << path
       end
     end
 
-    text_files
+    @text_files
   end
 
   def libtool_files
```
There's probably a better way to do this, since it would be useful to also cache the fact that `text_files` is genuinely empty, which this doesn't do.

danielnachun replied:
Yeah I was starting to think as we were discussing this that `text_files` should be able to cache its output.  Up until now it's only used internally by `replace_text_in_files`, so it's not called repeatedly.  

The difficulty as you said is that if `text_files` is actually empty, it would still end up trying to scan the keg every time you call it, thinking it hadn't been cached yet, when in reality there simply were no text files.  Maybe we need to initialize it to something other than `[]`, but which it would never sensibly have as a value, and return `@text_files` if doesn't have this other value.

carlocab replied:
Actually it's probably enough for the first line to be
```ruby
return @text_files if @textfiles.present?
```

MikeMcQuaid replied:
Agreed that the version you pitched is easier @danielnachun, I was wrong and @carlocab was right here 😁 

> `return @text_files if @textfiles.present?`

note that this won't cache `nil` or empty arrays.

danielnachun replied:
Maybe we need to have another variable `@no_text_files` initialized to `nil` that we set to `true` if the size of `text_files` is 0, and then the first two lines would be 
```
return @text_files if @text_files.present?
return [] if @no_text_files || !which("file") || !which("xargs")
```

MikeMcQuaid replied:
Nah, that's not necessary. If we want to cache `nil`: 

`return @text_files if defined?(@ text_files)`

If we want to cache an empty array:

`return @text_files if text_files`

If we want to only cache non-empty arrays:

`return @text_files if text_files.present?`

danielnachun replied:
Thanks that's super helpful. I believe we only need to be able to cache empty arrays as `text_files` should be `[]` if nothing is found and not `nil`.  I think I have an idea of how to verify `text_files` is caching within a unit test, and will try to have this cover both the empty case and the non-empty case when I open that smaller PR.

danielnachun replied:
I was going back over this in preparation to open a PR, and I realized that the only reason we even needed to use `text_files` was to get rid of sharballs.  However, we could just as easily do this with the following code:
```ruby
keg.each_unique_file_matching(old_prefix) do |file|
      next if !file.binary_file? # skip files which do not contain null bytes
      next if file.text_executable? # skip sharballs

   # binary patching code
end
```
Whether we use `text_files` or the above code, we still only iterate through the files in the keg once, and this way is much simpler and more self-contained.

I'm planning to totally refactor `text_files` and `replace_text_in_files` as part of the next major PR for this to update the bottle spec to store the names of all binaries with prefix matches, and to make sure we're doing prefix replacement in all text files, not the just the limited subset currently returned by `text_files`.  As part of that refactoring I think we'll be able to greatly simplify how `replace_text_in_files` and `text_files` will probably be unneeded.

MikeMcQuaid replied:
Sounds good.

danielnachun commented:
There's a lot of great feedback here so far which has identified a few things that I think may warrant their own smaller PRs first.  That seems to be a better strategy than try to cram too many big changes into one PR.  These would be the PRs:
1) In keg_relocate.rb, rename `apply_ad_hoc_signature` to `codesign_patched_binary`, and create an OS generic method that does nothing, which is overridden on macOS.  Note that we do use `apply_ad_hoc_signature` in contexts outside this PR but we're just renaming a method so it's not hard to change those other occurrences.
2) In `keg_relocate.rb`, revert `each_unique_file_matching` to its original form, and replace `each_unique_binary_file` with a much simpler `binary_file?` method that takes a single file as an argument and scans if for null bytes.  Ironically this means I broke `each_unique_file_matching` for nothing, but now I'm more comfortable with unit tests so it wasn't all bad!
3) Create a `cellar_from_tag` method in `software_spec.rb` to get the cellar from a tag, and refactor `compatible_locations?` to use this.

I know we've been working on this for a long time but given the scope of what is being implemented here I think it's worth doing this right!

MikeMcQuaid commented:
> There's a lot of great feedback here so far which has identified a few things that I think may warrant their own smaller PRs first. That seems to be a better strategy than try to cram too many big changes into one PR.

Agreed 👍🏻 

> I know we've been working on this for a long time but given the scope of what is being implemented here I think it's worth doing this right!

Agreed and these changes all sound good. This code is a bit gnarly and it's nice to see it being left in a better state than it was at the beginning, great work @danielnachun!

danielnachun commented:
I've now pushed a major update which has addressed most of the feedback so far and things are much cleaner now.  Here's a summary:

1) The `relocate_build_prefix` method now just uses a single loop over files with prefix matches, skipping files which are not binary or are sharballs. 
2) In `relocate_build_prefix`, the string splitting by null bytes no longer discards empty strings at the end, so we only have to pad the build prefixes and don't have to pad at the end of the file anymore.
3) I had to add back the generic `codesign_patched_binary` to `keg.rb` rather than `keg_relocate.rb`, because the macOS method is defined in `os/mac/keg.rg`.

The only thing that I think is still a little rough is how we handle deciding if a bottle needs prefix location.  The way the logic works is that a bottle will only be poured if one of the following conditions is met:
1) The bottle cellar is `:any` or `:any_skip_relocation`.
2) The user's cellar is the same as the bottle cellar.
3) The user's cellar is of equal or shorter length to the bottle cellar.

If a bottle is being poured, we still have to determine if the bottle needs prefix patching.  Prefix patching will be skipped if either of the following is true:
1) The bottle cellar is `:any` or `:any_skip_relocation`.
2) The user's cellar is the same as the bottle cellar.

The set of conditions in which we skip prefix patching is a strict subset of the set of conditions for us to decide if a bottle should be poured at all, but must be evaluated afterward.  As I've currently implemented it, the checks for whether the prefix should be patched are being done in `pour` in `formula_installer.rb`.  

We could move these checks into their own method, but that would bring us back to the original issue of having a somewhat redundant method with `compatible_locations?`.  Alternatively we could move the logic into `relocate_build_prefix` itself, though it may require a bit of refactoring.  I'm open to suggestions on the best way to implement this.

carlocab commented on Library/Homebrew/keg_relocate.rb:
```diff
@@ -165,6 +165,45 @@ def replace_text_in_files(relocation, files: nil)
+      file.ensure_writable do
+        binary = File.binread file
+        odebug "Replacing build prefix in: #{file}"
+        binary_strings = binary.split(/#{NULL_BYTE}/o, -1)
```
Do you use `NULL_BYTE` elsewhere? If not you may just want to make that constant a regex directly instead of into a string you interpolate into a `//o`.

danielnachun replied:
Unfortunately we do have to use `NULL_BYTE` as a string twice later in the method.  I suppose we could have a `NULL_BYTE_REGEX` to store the regex as a constant, but we only use it once so I'm not sure if that is necessary.

MikeMcQuaid reviewed (approved):
Looks good, nice work @danielnachun! Happy to merge when you're happy with this.

danielnachun commented:
I'm going to merge now given that we have much better testing coverage of this code and the scope of the PR is a lot smaller, so I am more confident now this won't break anything.

Bo98 commented on Library/Homebrew/keg.rb:
```diff
@@ -529,6 +529,8 @@ def binary_executable_or_library_files
     elf_files
   end
 
+  def codesign_patched_binary(_binary_file); end
```
This reintroduces the codesigning regression: https://github.com/Homebrew/brew/pull/12960
